Include both CSSLint and stylelint

Mottie opened the pull request:
As mentioned in #149, the stylelint bundle file is about 3.25 Mb in size. I'm not sure if that makes this a reasonable change, but I went ahead and did it.

If this PR is accepted, all the locale files will need to be updated in transifex to remove reference to CSSLint.

tophf commented:
Those who just install styles don't need a better linter and the significant increase of the extension.  Those authors who don't use advanced features provided by StyleLint would not enjoy the increase in the time needed to load and initialize the editor. The solution can be creating an additional editor extension so that Stylus would use it for its edit action.

tophf commented:
On the other hand, the zip increase is only 500kB -> 1MB, which is tolerable. And the increased load/parse times may not be an issue for those who absolutely need StyleLint. So we can still use this PR if you add an option to use CSSLint or StyleLint in the editor. The corresponding script(s) elements should be added based on the option value. And of course the editor should be able to use both inters.

tophf commented:
Ah, since there are CodeMirror addons that depend on linter script the entire script loading process should be programmatic, I guess, something like [this](https://github.com/tophf/stylish-chrome/blob/single-editor/editv2/editv2.js#L85)

narcolepticinsomniac commented:
Tolerable is a step up from unacceptable. The size increase is pretty drastic for something that'd only end up as an option though, no? I don't see how it's worth it if we're not committed to making it the default. Whether that's worth doing is apparently debatable. How noticeable are any speed issues? I tested briefly, and didn't see anything too out of whack. 

I see a few necessary overwrites right off the bat with StyeLint. Whatever is used will need tweaks, but with CSSlint, we're basically forking an abandoned project to maintain locally, whereas StyleLint can be expected to be generally maintained, correct?

Ultimately, you deal with most of the CSSLint bullshit tophf, so it's your call, but I'd think that a size increase this drastic would only be warranted for integral, default functionality, not an optional feature which'll never be utilized by the majority of users.

Mottie commented:
Stylelint is more actively maintained and also supports a lot of the newer css changes like variables and it'll work with the stylus preprocessor (for issue #134) out of the box. I know the bundle file is huge, but I think with a bit of work removing unnecessary code, we could bring that size down. The problem would be duplicating the modifications on updated versions of Stylelint.

Also, stylelint doesn't include a worker like CSSLint did, so there may be some performance differences.

tophf commented:
Well, the majority of users don't even open the editor hence my initial reluctance... As for the performance, my i7 PC or yours are probably faster than the average so I'm judging by the numbers in devtools profiler: StyleLint take as much time to load as the editor itself, linting is several times slower compared to CSSLint. With large styles and slower computers this difference will be noticeable.

tophf commented:
Well, the majority of users don't even open the editor hence my initial reluctance... As for the performance, my i7 PC or yours are probably faster than the average so I'm judging by the numbers in devtools profiler: StyleLint take as much time to load as the editor itself, linting is several times slower compared to CSSLint. With large styles and slower computers this difference will be noticeable.

tophf commented:
CSSLint doesn't use service workers, it's just a trivial code cosmetic that makes the script easier to include in a service worker. The performance difference is inherent because CSSLint is much more simple.

Mottie commented:
Hmm, maybe we could have a toggle to enable/disable active linting, and a button to check the linting on click. I'll look at it some more tomorrow.

tophf commented:
For advanced authoring it's better to implement the external editor support. Currently the development of that feature is stalled for unknown reason.

narcolepticinsomniac commented:
If we do end up switching to StyleLint as the default, here's the immediately obvious bugs.

"Write a new style" opens to an "unexpected empty source", non-typical selectors without `.` or `#` throw errors, errors are thrown pretty randomly for "unexpected longhand and shorthand" for valid code (see Github Dark), and `:increment` and `decrement` throw errors in scrollbar code.

Examples:

```
g-snapping-carousel {
    color: red;
}
::-webkit-scrollbar-button:horizontal:decrement {
    background-color: transparent;
}
```

Moz-format import/export are also broken.


I do remember back in the day, using Freestyler on a dual core, I had to disable the persistent validator. Not even sure what they were using, but the CPU usage was a bitch. The ability to revert to CSSLint should suffice, but options to completely disable, or only validate on save, might be useful to some. 

Pardon the ignorance, as always, but how efficient is the persistent validator in general? Is it a blunt instrument, constantly parsing the entire code, or is it only checking newly added code that hasn't already been validated? Even then, I would think it could only check for a few errors until matching brackets and then run the gambit. Not suggesting we could or should reinvent the validator, but I'm curious how resource intensive it is needlessly.

tophf commented:
All linters validate the entire style. I've been thinking today about limiting the area, but it's not trivial so I've postponed the thought until the editor is rewritten.

Mottie commented:
> "Write a new style" opens to an "unexpected empty source" error

I can change the config settings to ignore this. This would also mean modifying the "block-no-empty" rule (e.g. `.test {}`), but maybe it's better to show a warning instead of an error for this one?

> non-typical selectors without . or # throw errors

This is the typical behavior of the linter... it's hard to guess what selectors someone will use. These can be changed to warnings instead of errors.

> Errors are thrown pretty randomly for "unexpected longhand/shorthand" for valid code (see Github Dark)

I was testing this update with both GitHub-Dark and Stackoverflow-Dark. The "unexpected longhand/shorthand" issues are actually valid in this case. I could also change these to be warnings.

> :increment and :decrement throw errors in scrollbar code.

If I paste the scrollbar code into the [stylelint demo](https://stylelint.io/demo/), these errors pop up, so it's an issue with stylelint which is interesting since it appears to actually be [included in their keyword list](https://github.com/stylelint/stylelint/blob/f6fbad37e1ffa84aed4a996b3829b4275e7320c5/lib/reference/keywordSets.js#L276-L277). I opened an issue - https://github.com/stylelint/stylelint/issues/2810.

Mottie commented:
> how efficient is the persistent validator in general? Is it a blunt instrument, constantly parsing the entire code, or is it only checking newly added code that hasn't already been validated? Even then, I would think it could only check for a few errors until matching brackets and then run the gambit. Not suggesting we could or should reinvent the validator, but I'm curious how resource intensive it is needlessly.

I didn't extensively look at the `edit.js` code, but it appears that linting is updated through CodeMirror's set delay value (`cm.state.lint.options.delay`; which I have no idea where it's set) + 100 ms. I set an initial delay of 200 ms when the editor is first opened because it wasn't showing the issue panel at all without a delay.

It is pretty much a "blunt instrument" otherwise in that it checks the entire *section* on change; so it isn't too bad when I edit Stackoverflow-Dark with 40+ sections. It isn't too bad when I edit GitHub-Dark either... It doesn't seem that slow on my system, although I do have a older i7 quad core.

> it'd be reasonable to add a config dialog to choose the rules using checkboxes (and, ideally, with an import of a standard StyleLint config format).

@tophf Do you mean add settings for [standard rules](https://github.com/Mottie/stylus/blob/9473ed6ad1e707d1a5385a66cd326a9c9167d289/vendor-overwrites/codemirror/addon/lint/stylelint-config.js) or stylistic rules? or both? Maybe allowing the user to copy/paste their config JSON file into a separate textarea would be easier?

tophf commented:
I mean whatever makes sense. I don't use StyleLint but I know it has lots of rules which obviously won't suit everyone. In CSSLint we've enabled just four rules. Not sure the other ones are useful but maybe I'll add a config dialog for CSSLint as well. Someday.

Mottie commented:
@tophf I've been working on a dropdown select on the edit.html page, I can't seem to get the (dynamically) user selected linter to update... When the dropdown is modified, the `acmeEventListener` in `edit.js` is called. I added the following entry inside the `switch` function:

```js
case 'linter':
  if (value !== null && (editors.lastActive || editors[0])) {
    if (prefs.get(el.id) !== value) {
      prefs.set(el.id, value || 'csslint');
    }
    CodeMirror.signal(editors.lastActive || editors[0], "change");
    // save();
  }
  break;
```

If I signal a change on `editors.lastActive || editors[0]`, I see these error when switching the linter:

> foldgutter.js:107 Uncaught TypeError: Cannot read property 'state' of undefined
  at onChange (foldgutter.js:107)
  at Function.signal (codemirror.js:1175)
  at HTMLElement.acmeEventListener (edit.js:357)

The call to `save()` is supposed to work ([ref](https://stackoverflow.com/a/22812154/145346)), but it doesn't appear to update the issues container properly. Only after css content modification does the selected linter correctly report issues.

tophf commented:
I'd expect you would do something similar to `case 'autocompleteOnTyping'`, but anyway, I hope you can figure it out yourself as I'm not an infallible expert on CodeMirror/whatever.

tophf commented:
What about "tree shaking" to make the bundle smaller? I find it hard to believe that all that code is needed.

Mottie commented:
Well, I've tried webpack, browserify and rollup. The first two methods always seems to result in files 3+Mb in size - like it wasn't tree shaking at all. And, I just can't seem to get rollup to include node module files. I'll keep trying.

Mottie commented:
Ok, I think it's ready for review now... some notes:

* I have yet to produce an optimized (tree shaking) version of stylelint
  * Currently includes both a full size (3.25Mb) and a `.min` (1.4Mb - only whitespace & comments removed) version.
  * I seem to remember the reviewers at Chrome asking for un-minified files, so it's likely that neither of those files will be the final version.
  * I'll set up a repo with my current attempt at tree-shaking stylelint soon - then we can find & fix the problem and have a version that can be easily updated.
* A popup allowing modification of stylelint options has been added:
  * There is no validation, other than valid JSON on it. Will there need to be more validation, like the checking of valid rules?
  * Should these setting be set up with an instance of CodeMirror? If so, should I add the built-in `js` linter?
  * Adding all of the rules from [this demo](https://stylelint.io/demo/) causes some serious lag. I might add an update to use a generator to build the issue list in the header.
* All linting functions were split out from the `edit.js` and moved into `lint.js` file in the `edit` folder. Is that a good name, or too redundant & confusing?

tophf commented:
CodeMirror instance in a popup function like the one used for import-export, yes. And js linter addon (IIRC it's tiny), dynamically loaded of course just like beautifier (script loader may be refactored/promisified). Minimized vendor code is okay at least in case of well-known libraries, also its parsing and compiling are usually noticeably faster. I'll review the code tomorrow.

Mottie commented on edit.html:
```diff
@@ -33,10 +34,15 @@
     <script src="vendor/codemirror/addon/edit/matchbrackets.js"></script>
 
-    <link rel="stylesheet" href="vendor/codemirror/addon/lint/lint.css" />
+    <!-- CSSLint -->
```
The stylesheet is part of CodeMirror's linting tooltips. I would think that should be loaded no matter which linter is selected. Also, since CSSLint is the default and small, should that be loaded by default? Or do you want it to be dynamically loaded?

tophf replied:
Well, you can add a third option "none" to the linters selector. It would be internally represented as an empty string in `prefs` and `<option value="" i18n-text="genericNoneLabel">` in html. And genericNoneLabel in _locales/en/messages.json.

Mottie replied:
What if I used the "genericDisabledLabel" like that used in the Highlight dropdown?

Mottie replied:
And should "None" be the default? Or keep it on CSSLint?

tophf replied:
* genericDisabledLabel is okay, I forgot about it.
* ideally, stylelint should be the default, but it's too soon to make it so, we and other users need to test it, you may be able to reduce the bundle etc.
* so, practically, we should preserve the historic behavior: keep csslint

Mottie replied:
Done

Mottie commented on _locales/bg/messages.json:
```diff
@@ -665,8 +665,8 @@
 	"issuesHelp": {
-		"message": "Проблеми, намерени от <a href='https://github.com/CSSLint/csslint' target='_blank'>CSSLint</a> при следните правила:",
-		"description": "Help popup message for the CSSLint issues block on the style edit page"
+		"message": "Проблеми, намерени от $link$ при следните правила:",
```
Done

Mottie commented on edit.html:
```diff
@@ -186,6 +192,15 @@ <h2 id="options-heading" i18n-text="optionsHeading"></h2>
+            <option value="csslint" selected>CSSLint</option>
+            <option value="stylelint">stylelint</option>
+          </select>
+          <svg id="stylelint-settings" class="svg-icon settings">
```
Done

tophf commented on edit/lint.js:
```diff
@@ -0,0 +1,253 @@
+}
+
+function setupStylelintSettingsEvents() {
+  let timer;
```
Err, my comment above is mostly wrong because the variable is used in the inner function. Well, you can still use the existing `debounce()` function. Also, check if it's possible to close the dialog with an error, reopen the rules config dialog within 3 seconds and still get the previously scheduled error message. I think it's possible and you need to rework the `$('#help-popup .error')` check. Maybe use an expando property: `$('#help-popup .error').hideTimer = setTimeout(.......)` and check/clear it before invoking showHelp.

tophf replied:
Er, in more detail:

1. invoke the config, type something wrong, see an error message
2. reinvoke the config immediately, type something wrong again, see an error message
3. the error message is closed sooner than 3 seconds elapse

If this is possible, then see my 2nd comment above.

Mottie replied:
Even with the `hideTimer` set on the error message as an expando the `setTimeout` was still firing. I ended up moving the timer expando to the content container and using `clearTimeout` when the popup closes.

Mottie commented on edit/edit.js:
```diff
@@ -1,5 +1,5 @@
 /* eslint brace-style: 0, operator-linebreak: 0 */
-/* global CodeMirror exports parserlib CSSLint */
+/* global CodeMirror exports css_beautify parserlib CSSLint initLint setLinter updateLintReport renderLintReport updateLinter */
```
Done

Mottie commented on edit/edit.js:
```diff
@@ -2052,6 +1909,7 @@ function onRuntimeMessage(request) {
       break;
     case 'styleDeleted':
       if (styleId && styleId === request.id) {
+        // eslint-disable-next-line no-empty-function
```
Done

Mottie commented on edit/lint.js:
```diff
@@ -0,0 +1,253 @@
+/* global CodeMirror CSSLint editors makeSectionVisible showHelp stylelintDefaultConfig BG */
```
Done

Mottie commented on edit/lint.js:
```diff
@@ -0,0 +1,253 @@
+  BG.chromeLocal.setValue('editorStylelintRules', rules);
+}
+
+function setLinter(name) {
```
Done

Mottie commented on edit/lint.js:
```diff
@@ -0,0 +1,253 @@
+function setupStylelintSettingsEvents() {
+  let timer;
+  $('#help-popup .save').addEventListener('click', () => {
+    try {
```
Done

Mottie commented on edit/lint.js:
```diff
@@ -0,0 +1,253 @@
+        }
+      }, 3000);
+    }
+    return false;
```
Done

Mottie commented on edit/lint.js:
```diff
@@ -0,0 +1,253 @@
+    const link = '<a target="_blank" href="https://stylelint.io/demo/">Stylelint</a>';
+    const text = JSON.stringify({rules: rules}, null, 2);
+    const content = `<textarea class="contents settings">${text}</textarea>
+      <p>${t('setStylelintLink', link)}</p>
```
Will `$element` replace `tHTML` then?

tophf replied:
Eventually we'll replace all tHTML usage along with the function itself. Meanwhile we use it only for translated messages, and $element for the new DOM-building code.

Mottie replied:
Done

Mottie commented on vendor-overwrites/codemirror/addon/lint/stylelint-config.js:
```diff
@@ -0,0 +1,166 @@
+  rules: {
+    // ** recommended rules **
+    // ref: https://github.com/stylelint/stylelint-config-recommended/blob/master/index.js
+    'at-rule-no-unknown': [ true, { 'severity': 'warning' }],
```
Done

Mottie commented on vendor-overwrites/codemirror/addon/lint/linter.js:
```diff
@@ -0,0 +1,102 @@
+// CodeMirror, copyright (c) by Marijn Haverbeke and others
```
Done

Mottie commented:
I'll get to the rest of the changes when I get more time later today.

Mottie commented:
@tophf Here is my attempt at using rollup to [bundle stylelint](https://github.com/Mottie/stylelint-bundler)... the readme includes details on why I haven't got it working yet.

tophf commented:
* As for the bundle size, I thought it might be legit because the bundle includes implementation for all rules.
* The config icon should not be shown for CSSLint or you need to implement the actual config routine, which is currently hardcoded in vendor-overwrites/codemirror/addon/lint/css-lint.js to 5 rules.

tophf commented:
Ah, and please rebase on current master branch, then reupload the PR via `git push -f`

tophf commented:
And remove the unminified bundle, we'll risk it.

Mottie commented on edit/edit.js:
```diff
@@ -1314,7 +1185,8 @@ function initWithStyle({style, codeIsUpdated}) {
       cm.setOption('lint', CodeMirror.defaults.lint);
-      updateLintReport(cm, 0);
+      // update lint issue table after a short delay
+      updateLintReport(cm, 200);
```
Removed. This no longer appears to be an issue.

Mottie commented:
All done except for one remaining issue...

When the stylelint settings panel opens it used to look like it was working because the `css` linter was applied. Now that I have the `json-lint.js` and `jsonlint` library loaded. The syntax highlighting and linting doesn't appear to be working. This appears to be the correct json mode code:

```js
cm.setOption('mode', 'application/json');
cm.setOption('lint', 'json');
```

Can you tell if I'm missing something?

tophf commented:
To enable syntax highlight the `mode` should be `javascript` and you need to load the actual file: [mode/javascript/**javascript.js**](https://github.com/codemirror/CodeMirror/blob/7bd58c0c5c9212e68aa3558c785ec39658b6cc31/mode/javascript/javascript.js) (I'm linking the exact revision of CM we're currently using).

tophf commented:
Got an error in the background console, can't say if it's due to reloading or a miss on my part, but please investigate:

```
Uncaught (in promise) ReferenceError: stylelintDefaultConfig is not defined
    at setStylelintRules (edit/lint.js:20)
    at BG.chromeLocal.getValue.then.rules (edit/lint.js:15)
```

Mottie commented:
I can't seem to duplicate that error. The `stylelint-config.js` file (containing `stylelintDefaultConfig`) is loaded along with the other lint files, so it should be defined.

tophf commented:
It happens when chrome.storage.local doesn't have `editorStylelintRules` and you need to look in the background page console. The problem is that BG.chromeLocal.getValue runs in the background page and it returns a promise that also runs in the background page where stylelintDefaultConfig is not defined so there are two solutions: use chrome.storage.local.get/set directly or move chromeLocal to messaging.js and declare it in eslintrc.

Mottie commented:
Ok, *now* everything should be ready for review 😁

tophf commented:
Uhm wait, now the default rule list is empty on first install:

```js
{
  "rules": ""
}
```

Clicking "save" of course produces an "invalid json" error status because `rules` should be an array.
Deleting everything and saving can't get past the same error.
Only by specifying `[]` instead of `""` I can make it save an empty valid list so reopening the config finally displays the default list.

This is why I suggested those two solutions.

Mottie commented:
DOH... ok try it now. I've been avoiding the move of the storage function and only making changes in the `lint.js` file.

tophf commented:
Yep, now it works.

tophf commented on edit/edit.js:
```diff
@@ -508,7 +517,7 @@ window.onbeforeunload = () => {
   }
-  updateLintReport(null, 0);
+  updateLintReportIfEnabled(null, 0);
   return confirm(t('styleChangesNotSaved'));
```
js dialogs may be blocked by user setting, can you switch to `messageBox()` we use in manage.js for delete confirmation?

tophf replied:
er, disregard the above comment, it's about the old code, we'll handle it in the future

Mottie replied:
Hmm, I think the confirmation dialog is actually built-into the browser... if I type in `window.history.back();` in the dev tools console, I see the same confirmation popup.

Mottie replied:
Also, at least for me, `useHistoryBack` is always `false`.. so the code in `goBackToManage` function isn't ever called.

tophf replied:
useHistoryBack is true when there's a manage page to go back to: open the manager, click a style to edit it, then it'll be true.

Mottie replied:
LOL, ok... I've been kind of stuck on this one anyway... because `messageBox` returns a promise and `onbeforeunload` is expecting a boolean, it gets a lot more complicated.

tophf commented on edit/edit.js:
```diff
@@ -1011,145 +1019,11 @@ function getEditorInSight(nearbyElement) {
-    script.src = 'vendor-overwrites/beautify/beautify-css-mod.js';
-    script.onload = doBeautify;
-  }
+  const script = document.head.appendChild(document.createElement('script'));
```
Since you modify it anyway what about switch to the new script loader?

tophf commented on edit/lint.js:
```diff
@@ -0,0 +1,319 @@
+    editors.forEach(cm => {
+      // set lint to "null" to disable
+      cm.setOption('lint', options);
+      cm.refresh(); // enabling/disabling linting changes the gutter width
```
long line comment => should precede the line
care to check if there's a eslint rule for this and add it?

Mottie replied:
There is only a [`no-inline-comments` rule](https://eslint.org/docs/rules/no-inline-comments).  I'll go ahead and enable it and fix any issues that pop up.

tophf commented on edit/lint.js:
```diff
@@ -0,0 +1,319 @@
+
+  // touch devices don't have onHover events so the element we'll be toggled via clicking (touching)
+  if ('ontouchstart' in document.body) {
+    document.querySelector('#lint h2').addEventListener('click', toggleLintReport);
```
replace all document.querySelector with `$` and element.querySelector with `$(selector, element)` - at least in the new code.

tophf commented on edit/lint.js:
```diff
@@ -0,0 +1,319 @@
+  contents.insertBefore(popup.codebox.display.wrapper, contents.firstElementChild);
+  popup.codebox.focus();
+  popup.codebox.setValue(rules);
+  onDOMscripted(loadJSON).then(() => { setJSONMode(popup.codebox); });
```
{} braces aren't needed here

tophf commented on edit/lint.js:
```diff
@@ -0,0 +1,319 @@
+    // inject css
+    injectCSS('vendor/codemirror/addon/lint/lint.css');
+    // load CodeMirror lint code
+    scripts = scripts.concat([
```
```
scripts.push(
  'vendor/codemirror/addon/lint/lint.js',
  'vendor-overwrites/codemirror/addon/lint/css-lint.js'
);
```
?

Mottie replied:
LOL, I never realized that you can `push` more than one item <img title=":_ derp :" src="https://cloud.githubusercontent.com/assets/136959/13559387/49e2ecce-e3d9-11e5-8a16-73cbfdb7230e.gif">

tophf commented on edit/lint.js:
```diff
@@ -0,0 +1,319 @@
+  if (name === 'csslint' && !window.CSSLint) {
+    scripts.push('vendor/csslint/csslint-worker.js');
+  } else if (name === 'stylelint' && !window.stylelint) {
+    scripts = scripts.concat([
```
same

tophf commented on vendor-overwrites/codemirror/addon/lint/css-lint.js:
```diff
@@ -15,47 +15,85 @@
-  var found = [];
-  if (!window.CSSLint) return found;
+CodeMirror.registerHelper("lint", "csslint", function(text) {
+  let found = [];
```
ideally, it should be `const` as per our eslint rule, same for the occurrence below; however, this file has many other issues which you can see by removing vendor-overwrites/ from .eslintignore, so maybe you can fix them as well since you decided to reformat it but that's for you to decide.

tophf commented on vendor-overwrites/codemirror/addon/lint/css-lint.js:
```diff
@@ -15,47 +15,85 @@
+        }
+      }).then(output => {
+        const warnings = output.results.length ? output.results[0].warnings : [],
+          len = warnings.length;
```
a separate `const` for a variable with initializer should be used as per our eslint config, but see my comment above

tophf commented on vendor-overwrites/codemirror/addon/lint/stylelint-config.js:
```diff
@@ -0,0 +1,165 @@
+window.stylelintDefaultConfig = (defaultSeverity => ({
```
this file doesn't belong to codemirror so it shouldn't be in vendor-overwrites, move it edit/ or js/ depending on whether it's used somewhere besides the editor

tophf commented on edit/lint.js:
```diff
@@ -0,0 +1,323 @@
+    return;
+  }
+  if (delay > 0) {
+    setTimeout(cm => { cm.performLint(); update(cm); }, delay, cm);
```
violates our `brace-style` eslint rule: multiple statements should be on multiple lines

Mottie commented:
> Do you know which stylelint rule shows errors for invalid property values

Stylelint is missing a bunch of rules still... I opened up [this issue](https://github.com/stylelint/stylelint/issues/2816) because it wasn't validating `rgb(a)` and `hsl(a)` properties.

tophf commented:
But not catching `display: flexx` seems too unreasonable, rather egregious. How can stylelint be useful with such a glaring omission?

tophf commented on js/dom.js:
```diff
@@ -64,6 +64,83 @@ function onDOMready() {
+  Object.assign(script, properties);
+  if (!properties.onload) {
+    return new Promise(resolve => {
+      script.onload = () => { script.onload = null; resolve(); };
```
violates our brace-style rule, should be multiple lines and yeah I know it's my old code :)

Mottie commented:
> How can stylelint be useful with such a glaring omission?

At least it's being maintained... we can update it once new rules have been added.

tophf commented on edit/edit.css:
```diff
@@ -24,6 +24,10 @@ body {
 }
+section {
+  box-sizing: border-box;
+  width: 100%;
```
Why is this required? It breaks the layout in narrow width media queries. Disabling this rule doesn't change anything.

Mottie replied:
This was needed before I added the code (`magicBuffer` variable in `lint.js`) to subtract out a bit of padding to prevent the vertical scrollbar from showing up. Removed now.

tophf commented on edit/edit.css:
```diff
@@ -440,6 +471,10 @@ body[data-match-highlight="selection"] .CodeMirror-selection-highlight-scrollbar
 }
 #lint td[role="message"] {
   text-align: left;
+  max-width: 140px;
```
Why? It's better to view the entire message on multiple lines at one glance.

tophf replied:
BTW avoid changing the historical layout/appearance which isn't definitely buggy without discussion.

tophf replied:
I think we should keep the old code that cuts the message by max length, and additionally set the full text in cell's `title`.

tophf replied:
Now I remember why we started to cut the text in js: the browser will lag like hell when a SUUUUUUUUPER long line has an error and it's quoted fully in the message. For example, data URL for background image. CSS ellipsis overflow seems to be inherently slow.

Mottie replied:
Got it... I'll get to this soon. I'm working my way down the list 😁

Mottie commented:
All changes have been made... well except the `confirm` dialog thing.

tophf commented on .eslintrc:
```diff
@@ -151,7 +151,7 @@ rules:
   no-implicit-globals: [0]
   no-implied-eval: [2]
-  no-inline-comments: [0]
+  no-inline-comments: [2]
```
Ahem, now we have 50+ errors in other js files... I guess it was either a bad idea or you gave up too soon.

Mottie replied:
LOL... I guess I have more to do.

tophf replied:
Damn, it actually was a bad idea: the usercss PR will have a lot of conflicts because of this. This is not what I was hoping for when I asked for eslint rule... Such global cosmetic changes should occur in the absence of other PRs. I think it's better to revert the rule and the changes to the unrelated code.

Mottie replied:
How about we change the rule and keep the changes I made so we don't have to do it again?

tophf replied:
Redoing the changes vs hard-to-resolve and easy-to-mess conflicts? I think it's the former.

Mottie replied:
Ok... I'll have to get to this a little later this afternoon.

tophf commented on edit/lint.js:
```diff
@@ -93,23 +93,23 @@ function updateLintReport(cm, delay) {
           const rule = linter === 'stylelint' ?
-            info.message.substring(info.message.lastIndexOf('('), info.message.length) :
+            info.message.substring(info.message.lastIndexOf('('), info.message.length).replace(/[()]/g, '') :
             / at line \d.+$/;
```
It seems all stylelint messages start with the obvious and noisy when repeated `Unexpected` word which I suggest stripping just like we strip the `at line` from csslint's output.

Mottie replied:
So are you saying strip the rule from the table, or from CodeMirror's tooltip?

tophf replied:
Not the rule but the word `Unexpected` because it's redundant. And in the tooltip too.

Mottie replied:
Oh ok... I think the "at line" in the CSSLint output no longer exists - some recent changes stopped using the `rule` regex, so I'll clean that all up.

Mottie commented:
Shouldn't `csslint-worker.js` also be moved into the `vendor-overwrites\csslint` folder since it's been heavily modified?

Mottie commented:
Also, are you cool with me adding this to the `csslint-worker.js` file in this PR?

```diff
@@ -3714,7 +3714,7 @@ var Properties = module.exports = {
     "pitch-range"                   : 1,
     "play-during"                   : 1,
     "pointer-events"                : "auto | none | visiblePainted | visibleFill | visibleStroke | visible | painted | fill | stroke | all",
-    "position"                      : "static | relative | absolute | fixed",
+    "position"                      : "static | relative | absolute | fixed | sticky",
     "presentation-level"            : 1,
     "punctuation-trim"              : 1,
```

tophf commented:
Yeah, move it.
position: sticky is okay too.

Now the hard part: someone should actually test this PR.

Mottie commented on edit/edit.js:
```diff
@@ -1461,8 +1461,14 @@ function fromMozillaFormat() {
+  }
+
+  function doImportWhenReady(target) {
+    const replaceOldStyle = target.name === 'import-replace';
```
I'm not sure what changing the `name` of the clicked button is supposed to do, but I left it in there...

Mottie commented:
<img title=":_ onion_head_exhausted :" src="https://cloud.githubusercontent.com/assets/136959/13371483/cfdb63fe-dcec-11e5-8387-490ee41e7aa1.gif"> *sigh*

Now I'm not so sure about stylelint... since I guess it's only about style, not content - see https://github.com/stylelint/stylelint/issues/2811#event-1213730324 - which I think is ridiculous; but, I was directed to https://github.com/csstree/csstree which is a much smaller library. I haven't looked that closely, but it looks good.

tophf commented:
Is there a plug-in to use csstree in stylelint to make the latter more useful?

Mottie commented:
> Is there a plug-in to use csstree in stylelint to make the latter more useful?

I haven't had a chance to look yet... about to head to the gym. I'll look when I'm done.

Mottie commented:
I didn't mess with the comment changes in (71e4f5567fa892b) the `edit.js` and `lint.js` files, but everything else has been reverted.

tophf commented:
BTW it might be possible to reuse the reverted commit later via git cherry-picking. The main difference is that it'll be you dealing with conflicts :-)

Mottie commented:
It looks like stylelint plans on using [csstree as a plugin](https://github.com/stylelint/stylelint/blob/master/VISION.md#possible-errors)... but nothing has been implemented yet.

Mottie commented:
Well, I think it's too early to adapt csstree.

I can enter this entire block of invalid css into [their validator](https://csstree.github.io/docs/validator.html) and it thinks it looks good:

```css
body {
  d: huh;
  r: wha;
  x: test;
  y: whoa;
}

a {
  color: rgb(200%, 20%); /* min 3 definition */
  color: rgb(0, 0, 0, 0, 0, 0, 0); /* max 3 definitions */
  color: rgba(0); /* min 4 definitions */
  color: rgba(0, 0, 0, 0, 0, 0, 0); /* max 4 definitions */
  color: rgba(0, 0, 0, 100%); /* alpha channel can not be a percentage */
}

b {
  color: hsl(0); /* min 3 definitions */
  color: hsl(100%); /* hue can not be a percentage (number, deg, rad or turn) */
  color: hsl(0, 0, 0, 0, 0, 0, 0); /* max 3 definitions; */
  color: hsl(0, 50, 50); /* saturation & lightness must be percentages */
  color: hsla(0); /* min 4 definitions */
  color: hsla(0, 0, 0, 0, 0, 0, 0); /* max 4 definitions */
  color: hsla(0, 0%, 0%, 0%); /* alpha channel can not be a percentage */
}

/* Firefox v52+ only formats */
div {
  color: rgb(255 0 153 / 1);
  color: rgb(255 0 153 / 100%);
  color: hsl(270 60% 50% / .15);
  color: hsl(270 60% 50% / 15%);
  color: hsl(270deg, 60%, 70%);
  color: hsl(4.71239rad, 60%, 70%);
  color: hsl(.75turn, 60%, 70%);
}
```

tophf commented on js/dom.js:
```diff
@@ -14,7 +14,7 @@ for (const type of [NodeList, NamedNodeMap, HTMLCollection, HTMLAllCollection])
   // display a full text tooltip on buttons with ellipsis overflow and no inherent title
   const addTooltipsToEllipsized = () => {
-    for (const btn of document.getElementsByTagName('button')) {
+    for (const btn of $$('button')) {
```
This code is executed on resize event so it's better to keep getElementsByTagName as it's much-much-much faster than querySelectorAll used in $$.

tophf commented on js/localization.js:
```diff
@@ -30,7 +30,7 @@ function tHTML(html, tag) {
     const body = t.DOMParser.parseFromString(html, 'text/html').body;
     if (html.includes('i18n-')) {
-      tNodeList(body.getElementsByTagName('*'));
+      tNodeList($$('*', body));
```
This code is executed during page load so it's better to keep getElementsByTagName as it's much-much-much faster than querySelectorAll used in $$.

tophf commented on js/localization.js:
```diff
@@ -115,7 +115,7 @@ function tDocLoader() {
 
   // localize HEAD
-  tNodeList(document.getElementsByTagName('*'));
+  tNodeList($$('*'));
```
same

tophf commented on manage/filters.js:
```diff
@@ -293,7 +293,7 @@ function showFiltersStats({immediately} = {}) {
   $('#filters').classList.toggle('active', filtersSelector.hide !== '');
   const numTotal = BG.cachedStyles.list.length;
-  const numHidden = installed.getElementsByClassName('entry hidden').length;
+  const numHidden = $$('.entry.hidden', installed).length;
```
This code may be executed a lot while the user types something in the search input field. Even though this is a rare scenario overall, but it's better to keep getElementsByClassName as it's much-much-much faster than querySelectorAll used in $$.

tophf commented on edit/lint.js:
```diff
@@ -0,0 +1,330 @@
+  if (prefs.get('editor.linter') === 'csslint') {
+    header = t('issuesHelp', '<a href="https://github.com/CSSLint/csslint" target="_blank">CSSLint</a>');
+    list += CSSLint.getRules().map(rule =>
+      '<li><b>' + rule.name + '</b><br>' + rule.desc + '</li>'
```
```js
`<li><b><a href="${rule.url}">${rule.name}</a></b><br>${rule.desc}</li>`
```

Mottie commented:
I didn't check for useful CSSLint rules since it took so long to implement. I'm sure we can remove some of the old IE rules, but it's just as easy to leave them in there and allow the user to decide.

tophf commented:
So you've used a json format. Come to think of it, better than checkboxes (or dropdowns) I suggested and consistent with stylelint config as well as easier to backup/restore in some file/gist etc.

We're almost done I guess. The remaining issue is whether we can reduce the bundle size. I've checked the code manually and AFAICT these are the things that can be theoretically omitted:

* ~370kB of CanIUse data, is it really that useful? Maybe the rule that depends on that data isn't even enabled by default? Can't we (meaning you) download/compile/insert it at run time by adopting the corresponding stylelint plugin code (node_modules/browserslist/index.js)? It might be a button in stylelint config dialog "Download [caniuse](https://caniuse.com/) data".
* ~60kB of named colors alternatives as rgb() etc. I guess it can be useful though.

tophf commented:
It'd be reasonable to move linter config to chrome.storage.**sync**, but separate from `prefs`. The maximum item size in `sync` is [8192 bytes](https://developer.chrome.com/extensions/storage#property-sync-QUOTA_BYTES_PER_ITEM), with the overall limit at 100 kB. Our settings are ~1 kB. CSSLint config is ~1 kB, stylelint's could be much larger so you need to check JSON.stringify(rules) length and limit it at say 8000, display a warning.

There's no wrapper for the sync storage, so you can either copypaste `chromeLocal` or invoke chrome.storage.sync directly with callbacks.

Mottie commented:
I don't think Firefox supports `sync`... or at least it didn't the last time I looked.

tophf commented:
FF may not actually sync but the storage itself is supported since around FF52.

tophf commented:
Oh, we can raise the 8kB limit to ~40kB by compressing stringified json with lzstring and checking its length, which should be 8000 or less. The library is only a couple kB of super fast code (and [a faster mod](https://greasyfork.org/en/scripts/27531-lzstring-2xspeedup/code)), which produces 5x compression for stylelint-config.js contents.

tophf commented:
In case I wasn't clear, `prefs` is already using chrome.storage.sync (since 2 or so years ago), which is why I think lint config should also do that. An alternative is to add it to the exported styles json as an array item with a special id which won't break Stylish.

tophf commented:
Another issue is the default set of rules. I've clicked through some of the styles and I see some messages repeated multiple times, which I'd call petty complaints that obfuscate real issues under info noise:

* Longhand value ...... instead of ..... (shorthand-property-no-redundant-values)
* Expected shorthand property ....... (declaration-block-no-redundant-longhand-properties)
* Empty comment (comment-no-empty)

Do these really need to be enabled by default? Are there some popular configs for stylelint, preferably acknowledged by many CSS authors?

narcolepticinsomniac commented:
>Do these really need to be enabled by default?

Hell no. When I last checked, I brought this up. They were also mistakenly flagged as errors IMO, when they should only be warnings, given that they're valid code which could/should be optimized. Errors should be reserved for invalid/broken code, and I wouldn't even want warnings for the above mentioned "issues".

Mottie commented:
> move linter config to chrome.storage.sync

Did you want me to add the a new `chromeSync` function to `storage.js`, replace `chromeLocal` (because then it'll synchronize across platforms), or should I make the lint configuration storage independent?

> compressing stringified json with lzstring

Did you want this added to the `chromeLocal` (if we're keeping it), or just for because of conflicts, only add it to `chromeSync`?

> ~60kB of named colors alternatives as rgb() etc.

Removed and it reduced the file size by about 50kB.

> ~370kB of CanIUse data

This data is loaded by autoprefixer. And used by the following rules:

* `at-rule-no-vendor-prefix`
* `media-feature-name-no-vendor-prefix`
* `property-no-vendor-prefix`
* `selector-no-vendor-prefix`
* `value-no-vendor-prefix`

Which is interesting because none of these rules were included in their [online demo](https://stylelint.io/demo/) - where I copied all the rule's from originally.

I removed these 5 rules, commented out a bunch of other non-essential stuff, and removed white-space with Closure compiler and got the file down to 871 kB! 🎉- I'll add the details of the process in my [fork of stylelint](https://github.com/Mottie/stylelint/tree/mod) - mod branch.

tophf commented:
You don't like the idea of dynamically downloaded caniuse data? Well, we can add this feature in the future, I guess. Same for named colors. 

chromeLocal is used for discardable stuff, currently update log. In the future, if or when we add a fallback #169, it might be also used for style database in FF. I don't know if chromeSync wrapper is needed, it depends on usage frequency. Like maybe 5+ times.

tophf commented:
Ah, re lzstring, it makes sense for sync because it's the only storage that has such a draconian restriction on item size.

tophf commented:
re `special id`, I thought it'd be a good idea to add all settings to the style backup file, but I can do it in the future. Actually, I can switch to sync myself after merging. 

Let's advance with what we have now. Before merging this PR, I'd like you to fix the exception thrown when an supported rule is specified in the stylelint config dialog:

```
Uncaught (in promise) Error: Undefined rule at-rule-no-vendor-prefix
    at l.exports (vendor-overwrites/stylelint/stylelint-bundle.min.js:1600)
    at vendor-overwrites/stylelint/stylelint-bundle.min.js:1120
    at Array.forEach (<anonymous>)
    at m (vendor-overwrites/stylelint/stylelint-bundle.min.js:1120)
    at vendor-overwrites/stylelint/stylelint-bundle.min.js:1123
    at <anonymous>
```

It'd be acceptable if you can show a warning right on clicking the "Save" button in the config dialog like "The following rules were not applied due to blabla. This feature is not yet implemented in Stylus, we're working on it`. It'd be excellent if you can actually implement it now and dynamically download caniuse data (a messageBox should be shown with on OK/cancel) and enable the dependent rules, but not really necessary at this point.

tophf commented on _locales/en/messages.json:
```diff
@@ -481,6 +494,31 @@
+      }
+    }
+  },
+  "resetLinterRules": {
```
This is a totally generic label which we might reuse in the fure, rename it to "genereicResetLabel" and move it to the other generic labels.

Mottie replied:
Done

tophf commented on _locales/en/messages.json:
```diff
@@ -481,6 +494,31 @@
+  },
+  "resetLinterRules": {
+    "message": "Reset",
+    "description": "Reset Stylelint or CSSLint rules"
```
And the description should be copied from other generic labels e.g. "Used in various parts of UI to indicate that something may be reset to its original state"

tophf commented on edit/lint.js:
```diff
@@ -0,0 +1,378 @@
+    linter === 'stylelint'
+      ? 'editorStylelintRules'
+      : 'editorCSSLintRules'
+    ).then(rules => {
```
causes an eslint error; this single line should be unindented by 2 spaces

Mottie replied:
I've disabled the `indent-legacy` and `no-multi-spaces` eslint rules locally otherwise Atom pops up error messages while I type. I really should fix it or start using VS Code.

Mottie commented:
> You don't like the idea of dynamically downloaded caniuse data?

Oh, I misread that before... anyway, I was in the process of ripping code out and didn't think about messing around with autoprefixer. I'll take a look.

> fix the exception thrown when an supported rule is specified in the stylelint

LOL ouch... yeah pretty much anything you add that isn't a rule breaks stylelint. I'll look into it later today.

Mottie commented on msgbox/msgbox.js:
```diff
@@ -42,9 +42,11 @@ function messageBox({
-    setTimeout(messageBox.resolve, 0, value);
-    animateElement(messageBox.element, {className: 'fadeout', remove: true})
-      .then(unbindAndRemoveSelf);
+    if (messageBox.resolve) {
```
Without this check, I was getting errors due to `messageBox.resolve` being `null` - this was occurring after I hit escape to close the help-popup, but only after `messageBox` was previously visible and closed.

tophf replied:
It means the keydown listener was never detached so a better solution is to investigate why. Turns out there was a typo: `document.removeEventListener('keydown'` should be `window.`. Also, `Esc` key should only close messageBox when both are displayed. I've push a proper fix in master so please revert this change.

Mottie commented:
To me, it looks like supporting the 5 vendor-prefix rules isn't worth it:

* The [caniuse-lite](https://github.com/ben-eb/caniuse-lite) zipped database is 534 kB in size.
* [Autoprefixer](https://github.com/postcss/autoprefixer/) would need to load and process the database before it would work with the rules, but only if called by a rule function.
* If one of those 5 rules is in use, we would probably (I haven't checked) need to modify each rule to work with the asynchronous autoprefixer code.
* We would need to keep track of all the changes made to autoprefixer, separately from the [stylelint mod](https://github.com/Mottie/stylelint/tree/mod), in case it gets updated.

I'm disinclined to continue working on this unless there is a demand for those 5 rules.

tophf commented:
>the 5 vendor-prefix rules isn't worth it:

Agreed.
We'll do it someday, *probably*.

tophf commented on edit/lint.js:
```diff
@@ -0,0 +1,413 @@
+      }),
+      makeButton('save', 'styleSaveLabel'),
+      makeButton('reset', 'genericResetLabel'),
+      $element({
```
The config dialog has `Save` and `Reset` buttons so I think it should also have a `Cancel` button (i18n message is `confirmCancel`) in order to present all available actions in one place.

Mottie replied:
Do you think the reset button should have a confirmation popup? Or maybe move the button to the right side to minimize chances of accidental clicking?

tophf replied:
Hmm, you can add a tooltip "To undo accidental reset, simply press Ctrl-Z (or Cmd-Z) in the text box".

tophf commented on edit/lint.js:
```diff
@@ -0,0 +1,413 @@
+      setCSSLintRules();
+      rules = {rules: csslintDefaultRuleset};
+    }
+    popup.codebox.setValue(JSON.stringify(rules, null, 2));
```
The result of JSON.stringify for rules isn't easy to oversee at a glance, mainly because of `{"severity": "warning"}` occupying 3 lines.

Try tweaking it with something like `.replace(/,\n\s+\{\n\s+("severity":\s"\w+")\n\s+\}/gm, ', {$1}')`

tophf replied:
correction, `m` flag is not required in regexp.

Mottie replied:
What do you think of putting it all on one line, except for the extended options. Something like this:

```js
{
  "at-rule-no-unknown": [true, {"severity": "warning"}],
  "block-no-empty": [true, {"severity": "warning"}],
  "color-no-invalid-hex": [true, {"severity": "warning"}],
  "declaration-block-no-duplicate-properties": [true,
  {
    "ignore": ["consecutive-duplicates-with-different-values"],
    "severity": "warning"
  }],
  "declaration-block-no-shorthand-property-overrides": [true, {"severity": "warning"}],
  "font-family-no-duplicate-names": [true, {"severity": "warning"}],
  "function-calc-no-unspaced-operator": [true, {"severity": "warning"}],
  "function-linear-gradient-no-nonstandard-direction": [true, {"severity": "warning"}],
  "keyframe-declaration-no-important": [true, {"severity": "warning"}],
  "media-feature-name-no-unknown": [true, {"severity": "warning"}],
  "no-empty-source": false,
  "no-extra-semicolons": [true, {"severity": "warning"}],
  "no-invalid-double-slash-comments": [true, {"severity": "warning"}],
  "property-no-unknown": [true, {"severity": "warning"}],
  "selector-pseudo-class-no-unknown": [true, {"severity": "warning"}],
  "selector-pseudo-element-no-unknown": [true, {"severity": "warning"}],
  "selector-type-no-unknown": [true, {"severity": "warning"}],
  "string-no-newline": [true, {"severity": "warning"}],
  "unit-no-unknown": [true, {"severity": "warning"}],
  "comment-no-empty": false,
  "declaration-block-no-redundant-longhand-properties": false,
  "shorthand-property-no-redundant-values": false
}
```

tophf replied:
It was my initial idea but then I decided not seeing true/false at a predictable position is rather confusing.

tophf commented on edit/lint.js:
```diff
@@ -0,0 +1,413 @@
+        ? setStylelintRules(rules)
+        : setCSSLintRules(rules);
+    }
+    const rulesString = JSON.stringify({rules: rules}, null, 2);
```
same

tophf commented on edit/csslint-ruleset.js:
```diff
@@ -0,0 +1,49 @@
+ * CSSLint Ruleset values
+ * 0 = disabled; 1 = warning; 2 = error
+ */
+window.csslintDefaultRuleset = {
```
Why `ruleset`? Can't we use `rules` for all linters? If we can then all occurrences of "ruleset" should be replaced.

Mottie replied:
That's just what CSSLint calls them... rules (the actual definitions) are separate from ruleset (the state of the rule: disabled, warning or error).

tophf replied:
Ah, I see the reason, but for Stylus those are identical entities - linter config. Maybe `config` or `ruleConfig` etc.

tophf commented on _locales/en/messages.json:
```diff
@@ -135,6 +135,10 @@
     "message": "Double-click to maximize/restore the height",
     "description": "Tooltip for the resize grip in style editor"
   },
+  "cm_linter": {
```
Now cm_linter causes a conflict with master. Better move it to maintain sorting order within cm_* messages. 

We still have some messages that were added at random positions, but I fix those when doing something meaningful with the related code.

Mottie replied:
My OCD kicked in and I ended up sorting all the "cm_*" messages... I reined it in before sorting all of the entries.

Mottie replied:
Ahh derp.. you mean conflicts with the master branch... I need more caffeine.

tophf commented on edit/lint.js:
```diff
@@ -0,0 +1,424 @@
+  contents.insertBefore(popup.codebox.display.wrapper, contents.firstElementChild);
+  popup.codebox.focus();
+  popup.codebox.setValue(rules);
+  onDOMscripted(loadJSON).then(() => setJSONMode(popup.codebox));
```
Add `popup.codebox.clearHistory();` after codebox.setValue, otherwise it's possible to undo to an empty contents via Ctrl-Z. Leave reset as is - to make it possible to undo a reset.

tophf commented on _locales/en/messages.json:
```diff
@@ -331,11 +343,20 @@
+      }
+    }
+  },
+  "stylelintConfig": {
```
Simply "Configure rules" - this message is used for CSSLint too.

It'd be reasonable to group all linter related messages via `linter` prefix: linterConfigRules, linterAllRulesLink, linterRulesTitle, linterConfigParseError, linterConfigInvalidRulesError, linterCSSLintLegend

tophf replied:
er, linterConfigTitle

tophf commented:
Also wow, stylelint is superslow on big code sections ([example](https://userstyles.org/styles/31267/global-dark-style-changes-everything-to-dark)), we'll need to make it autostop on any keydown/mousedown event.

tophf commented:
Actually, I'm not sure it's a good idea to merge this until we implement autostop.

tophf commented:
If hacking into stylelint is too difficult, we can start with the following:

1. add a checkbox in the rule config for stylelint to `process only on saving/opening`.
2. measure how long it took to run linter and adjust the auto-linting delay to a 5x multiple. Measurement can be done via `performance.now()` as we do already in various parts of the code.

Mottie commented:
> If hacking into stylelint

Maybe we can use a generator inside or intermittent loops inside of the `css-lint.js` file.

I have other plans this weekend, so I don't know when I'll get a chance to work on it.

Mottie commented:
Well, plans changed now that's its raining outside LOL...

What if we have a global variable that is set while the user is actively typing...

```js
document.addEventListener('keyup', () => {
  clearTimeout(window.userTyping);
  window.userTyping = setTimeout(() => {
    window.userTyping = null;
  }, 500);
});
```

Then in the `css-lint.js` file we `return false`, or something, instead of calling `stylelint.lint()`.

I haven't tested the idea to see if it works, but we may have to completely disable the linter dynamically. Something like this:

```js
document.addEventListener('keyup', () => {
  if (prefs.get('editor.linter') === 'stylelint') {
    clearTimeout(window.userTyping);
    cm.setOption('lint', false);
    window.userTyping = setTimeout(() => {
      cm.setOption('lint', 'stylelint');
    }, 500);
  }
});
```

tophf commented:
Other event should be hooked too like mousedown and touchstart in case `'ontouchstart' in document.body`. Also, 500 may be too small or too big, which is why I suggested measuring the actual time.

Mottie commented:
I wasn't seeing the huge delay that you mentioned...

In Global Dark style

![2017-08-27 09_48_42-developer tools - chrome-extension___fndlfdibhmbclfpjmbokjpgnehgjdkgn_edit html_](https://user-images.githubusercontent.com/136959/29752143-2dbfe964-8b1e-11e7-801d-ce5eb1bbdc1b.png)

Or in the StackOverflow Dark style

![2017-08-27 09_57_43-developer tools - chrome-extension___fndlfdibhmbclfpjmbokjpgnehgjdkgn_edit html_](https://user-images.githubusercontent.com/136959/29752145-47c689e4-8b1e-11e7-807c-681b2009ca63.png)

tophf commented:
So, I can merge this now and rename stylelint config object myself, or will you do it? :-)

Mottie commented:
Working on it now.

Mottie commented:
Also, I've started encountering one minor issue... it's not saving the `editor.linter` setting after I switch back and forth to the manage.html page. I can't seem to find where that problem might have been introduced, but I've been working on other things and haven't had a chance to isolate it.

tophf commented:
Make sure you've reloaded the extension, maybe this issue is caused by switching to a different branch temporarily.

Mottie commented:
Ugh, this is really frustrating...

If I enter `prefs.set('editor.linter', 'stylelint')` in the console of the `edit.html` page, I see the pref is updated. Immediately after going to the `manage.html` page, I enter `prefs.get('editor.linter')` in the console and see that it's been reset to "csslint". Every time.

The opposite is also true... enter `prefs.set('editor.linter', 'stylelint')` on the `manage.html` page and upon switching to the `edit.html` page, it resets back to "csslint" (the log was added to the beginning of the `initCodeMirror` function).

 Any ideas?

tophf commented:
I'll look into it tomorrow. I think I've fixed a similar issue today in the master branch.

tophf commented:
Yes, it's the same issue. You've rebased on a buggy master.

tophf commented:
@Mottie, I've reworked the config-related things myself as it was too complicated to explain. And changed a few other trivial things. Turns out you've used `'null'` when no linter is active, which I thought pretty weird (what if we add a `null` linter? lol), so an empty string is used (IIRC this is what I suggested initially).

Mottie commented:
I think you're actually supposed to set the CodeMirror option to `null`, but the code was expecting a string in a few places and `'null'` also worked, so I left it 😛.

I'm glad it's finally merged! 🎉!

tophf commented:
It's supposed to be `false` and your code was setting it correctly in getForCodeMirror (in the ternary's alternative branch). I didn't change that.

tophf commented:
BTW, an easy instruction or a build script will be required on AMO to produce an exact copy of our minified bundle.

Mottie commented:
I included a markdown file linking to the repo...

https://github.com/openstyles/stylus/blob/master/vendor-overwrites/stylelint/stylelint-mod.md

tophf commented:
I know, but I don't see an easy instruction or a build script. I meant something that @schomery can include in a note when submitting to AMO.

Mottie commented:
https://github.com/Mottie/stylelint/tree/mod#stylelint-mod

tophf commented:
I know, but I'm afraid the AMO drones won't be able to follow that. They also don't have time. When I say an easy instruction I mean something that's really easy.

tophf commented:
A build script that performs all modifications would be ideal.

Mottie commented:
Ok, I'll see what I can do.

tophf commented:
1.1.4.1 is listed on AMO so apparently they've managed to follow your instructions.

Mottie commented:
Nice! Hmm, too bad the Chrome version still hasn't been approved 😞
